src/cache/cloud_fraction.jl

```diff
@@ -0,0 +1,159 @@
+import StaticArrays as SA
+import ClimaCore.RecursiveApply: rzero, ⊞, ⊠
+
+# TODO: write a test with scalars that are linear with z
+"""
+    Diagnose horizntal covariances based on vertical gradients
+    (i.e. taking turbulence production as the only term)
+"""
+function covariance_from_grad(coeff, ᶜmixing_length, ᶜ∇Φ, ᶜ∇Ψ)
+    return 2 * coeff * ᶜmixing_length^2 * dot(ᶜ∇Φ, ᶜ∇Ψ)
+end
+
+"""
+   Compute the grid scale cloud fraction based on sub-grid scale properties
+"""
+function set_cloud_fraction!(Y, p, ::DryModel)
+    FT = eltype(Y)
+    (; cloud_fraction) = p.precomputed
+    @. cloud_fraction = FT(0)
+end
+function set_cloud_fraction!(Y, p, ::Union{EquilMoistModel, NonEquilMoistModel})
+    (; SG_quad, params) = p
+
+    FT = eltype(params)
+    thermo_params = CAP.thermodynamics_params(params)
+    ᶜdz = Fields.Δz_field(axes(Y.c))
+    (; ᶜts, ᶜp, cloud_fraction) = p.precomputed
+
+    ᶜqₜ = p.scratch.ᶜtemp_scalar
+    ᶜθ = p.scratch.ᶜtemp_scalar_2
+    @. ᶜqₜ = TD.total_specific_humidity(thermo_params, ᶜts)
+    @. ᶜθ = TD.liquid_ice_pottemp(thermo_params, ᶜts)
+
+    coeff = FT(2.1) # TODO - move to parameters
+
+    @. cloud_fraction = quad_loop(
+        SG_quad,
+        ᶜp,
+        ᶜqₜ,
+        ᶜθ,
+        covariance_from_grad(
+            coeff,
+            ᶜdz, # TODO - replace dz with mixinglength when using EDMF SGS
+            Geometry.WVector(ᶜgradᵥ(ᶠinterp(ᶜqₜ))),
+            Geometry.WVector(ᶜgradᵥ(ᶠinterp(ᶜqₜ))),
+        ),
+        covariance_from_grad(
+            coeff,
+            ᶜdz,
+            Geometry.WVector(ᶜgradᵥ(ᶠinterp(ᶜθ))),
+            Geometry.WVector(ᶜgradᵥ(ᶠinterp(ᶜθ))),
+        ),
+        covariance_from_grad(
+            coeff,
+            ᶜdz,
+            Geometry.WVector(ᶜgradᵥ(ᶠinterp(ᶜθ))),
+            Geometry.WVector(ᶜgradᵥ(ᶠinterp(ᶜqₜ))),
+        ),
+        thermo_params,
+    )
+end
+
+"""
+    function quad_loop(SG_quad, p_c, qt_mean, θl_mean, qt′qt′, θl′θl′, θl′qt′, thermo_params)
+
+where:
+  - SG_quad is a struct containing information about quadrature type and order
+  - p_c is the atmospheic pressure
+  - qt_mean, θl_mean is the grid mean q_tot and liquid ice potential temperature
+  - qt′qt′, θl′θl′, θl′qt′ are the (co)variances of q_tot and liquid ice potential temperature
+  - thermo params are the thermodynamics parameters
+
+The function trnasforms and imposes additional limits on the quadrature points.
+It returns cloud fraction computed as a sum over quadrature points.
+"""
+function quad_loop(
+    SG_quad::SGSQuadrature,
+    p_c,
+    qt_mean,
+    θl_mean,
+    qt′qt′,
+    θl′θl′,
+    θl′qt′,
+    thermo_params,
+)
+    FT = eltype(qt_mean)
+
+    @assert SG_quad.quadrature_type isa GaussianQuad
+
+    sqrt2 = FT(sqrt(2))
+
+    # Epsilon defined per typical variable fluctuation
+    eps_q = FT(eps(FT)) * max(FT(eps(FT)), qt_mean)
+    eps_θ = FT(eps(FT))
+
+    # limit σ_q to prevent negative q_tot_hat
+    σ_q_lim = -qt_mean / (sqrt2 * SG_quad.a[1])
```

Comment:
Good point. I'll go over it and unify the notation. Since now it's become an internal diagnostics routine, could I stick with short names and just go with `q` and `θ` everywhere inside the quadrature loops? I'll make sure to document what those are